src/ILLink.Shared/DataFlow/ValueSet.cs

```diff
@@ -9,49 +13,161 @@ namespace ILLink.Shared.DataFlow
 	public readonly struct ValueSet<TValue> : IEquatable<ValueSet<TValue>>, IEnumerable<TValue>
 		where TValue : notnull
 	{
-		public readonly HashSet<TValue>? Values;
+		// Since we're going to do lot of type checks for this class a lot, it is much more efficient
+		// if the class is sealed (as then the runtime can do a simple method table pointer comparison)
+		sealed class EnumerableValues : HashSet<TValue>
+		{
+			public EnumerableValues (IEnumerable<TValue> values) : base (values) { }
+
+			public override int GetHashCode ()
+			{
+				int hashCode = 0;
+				foreach (var item in this)
+					hashCode = HashUtils.Combine (hashCode, item);
+				return hashCode;
+			}
+		}
+
+		public struct Enumerator : IEnumerator<TValue>, IDisposable, IEnumerator
+		{
+			readonly object? _value;
+			int _state;  // 0 before begining, 1 at item, 2 after end
+			readonly IEnumerator<TValue>? _enumerator;
+
+			internal Enumerator (object? values)
+			{
+				_state = 0;
+				if (values is EnumerableValues valuesSet) {
+					_enumerator = valuesSet.GetEnumerator ();
+					_value = default (TValue);
```

Comment:
Should this be `null`? Alternatively, would it be possible to store the `_enumerator` in the same `object?` field?

Reply:
It should be null - you're right.

We could store the enumerator in the same field, but it would require more casting, which is slower. In the `ValueSet` the memory footprint is important as we will store these values for a relatively long time and so they add up. On the other hand the `Enumerator` class should be VERY short lived. Basically always only stack allocated, so trading memory for CPU here feels like the better option.